cmd/gcp-controller-manager/kubelet_readonly_csr_approver.go

```diff
@@ -0,0 +1,327 @@
+			Status:  v1.ConditionTrue,
+		})
+	}
+	updateRecordMetric := csrmetrics.OutboundRPCStartRecorder("k8s.CertificateSigningRequests.updateApproval")
+	_, err := approver.ctx.client.CertificatesV1().CertificateSigningRequests().UpdateApproval(context.TODO(), csr.Name, csr, metav1.UpdateOptions{})
+	if err != nil {
+		klog.Errorf("error updating approval status for csr: %v", err)
+		updateRecordMetric(csrmetrics.OutboundRPCStatusError)
+		return fmt.Errorf("error updating approval status for csr: %v", err)
+	}
+	updateRecordMetric(csrmetrics.OutboundRPCStatusOK)
+	return nil
+}
+
+type readonlyResourceCsrValidator struct {
+	name          string
+	authFlowLabel string
+
+	// validate will validate the request and return approved or denied
+	// (false, err) will base on err type retry or not. if it is a non-retryable error then deny the request
+	// (true, nil) will continue to next validater. if it is the last one, will approve the request
+	// (false, nil) reject the request.
+	validate   readonlyResourceCSRValidateFunc
+	permission authorization.ResourceAttributes
+}
+
+func newReadonlyResourceCsrValidator() []readonlyResourceCsrValidator {
+	return []readonlyResourceCsrValidator{
+		{
+			name:          "csr usage validator",
+			authFlowLabel: "kubelet_readonly_csr_usage_match",
+			validate:      validateCSRUsage,
+		},
+		{
+			name:          "pod annotation validator",
+			authFlowLabel: "kubelet_readonly_pod_annotation_valid",
+			validate:      validatePodAnnotation,
+		},
+		{
+			name:          "rbac validator",
+			authFlowLabel: "kubelet_readonly_rbac_exist",
+			validate:      validateRbac,
+		},
+	}
+}
+
+func validateCSRUsage(content kubeletReadonlyCSRRequestContent) kubeletReadonlyCSRRequestResult {
+	csr := content.csr
+	if csr == nil || len(csr.Spec.Usages) == 0 {
+		klog.Errorf("csr usage is empty")
+		return kubeletReadonlyCSRRequestResult{
+			result:  false,
+			err:     nil,
+			message: "csr usage is empty",
+		}
+	}
+
+	usageMap := map[capi.KeyUsage]struct{}{}
+	for _, u := range kubeletReadonlyUsageAllowList {
+		usageMap[u] = struct{}{}
+	}
+
+	for _, usage := range csr.Spec.Usages {
+		if _, ok := usageMap[usage]; !ok {
+			allowListString, _ := json.Marshal(kubeletReadonlyUsageAllowList)
+			klog.Errorf("csr usage %v is not allowed, only allow any of %v", usage, string(allowListString))
+			return kubeletReadonlyCSRRequestResult{
+				result:  false,
+				err:     nil,
+				message: fmt.Sprintf("csr usage %v is not allowed, only allow any of %v", usage, string(allowListString)),
+			}
+		}
+	}
+	return kubeletReadonlyCSRRequestResult{
+		result:  true,
+		err:     nil,
+		message: "csr usage is valid",
+	}
+}
+
+func validatePodAnnotation(content kubeletReadonlyCSRRequestContent) kubeletReadonlyCSRRequestResult {
+	if !*isAutoPilot {
+		return kubeletReadonlyCSRRequestResult{
+			result:  true,
+			err:     nil,
+			message: fmt.Sprintf("Cluster is not an aotu pilot cluster, Annotation validation bypass"),
```

Comment:
typo: s/aotu/auto/, 
also, please verify the correct case/spelling the GKE Autopilot product name.